metalgpy/_sample.py

```diff
@@ -36,12 +38,104 @@ def list_type(list_exp):
+                 exp,
+                 distributions: dict=None,
+                 rng:np.random.RandomState=None):
         self.exp = exp
-        self.distributions = distributions
+        self.distributions = self._map_dist(distributions)
         self.rng = rng
+        self._cache = {}
+
+    def sample(self, size=None, to_nd_array=True):
+        # add a list to store samples
+        samples = []
+
+        # init a random state is rng
+        if self.rng is None:
+            self.rng = np.random.RandomState()
 
+        # check if the size is None
+        if size is None:
+            size = 1
 
+        # gather the list of choices
+        choices = self.exp.choices()
+
+        # get the required number of samples
+        for _ in range(size):
+            choice = self._sample(choices)
+
+            # store the sample
+            samples.append(choice)
+
+        # add a try catch block to capture returning elements to an nd-array
+        try:
+            if to_nd_array:
+                # convert the samples to an nd_array
+                samples = np.array(samples).reshape(len(samples), -1)
+        except ValueError:
+            # warn about invalid array structure
+            print("Cannot convert the sample to a numpy array")
+
+        # return the list in a numpy array
+        return samples
+
+    def _sample(self, choices):
+        # init a list to store samples
+        s = []
+
+        for var_id, var_exp in choices.items():
+            # check if the expression is Float
+            if isinstance(var_exp, Float):
+                # add samples
+                s.append(self.distributions[var_id].rvs(loc=var_exp._low, \
+                                           scale=var_exp._high - var_exp._low, \
+                                           size=1, random_state=self.rng))
+
+            # check if the expression is Int
+            if isinstance(var_exp, Int):
+                # add samples
+                s.append(self.distributions[var_id].rvs(low=var_exp._low, \
+                                           high=var_exp._high + 1, \
+                                           size=1, random_state=self.rng))
+
+            # check if the expression is List
+            if isinstance(var_exp, List):
+                # get idx of the samples to choice
+                var_exps_idx = self.rng.choice(var_exp._length(), size=1)
+
+                # add samples
+                s.append(var_exp._getitem(var_exps_idx))
+
+        # return the samples
+        return s
+
+    def _map_dist(self, distributions):
+        # create an empty dist map
+        dist_map = {}
+
+        # check if the distributions are passed
+        if distributions is None:
+            dist_map = {var_id: var_exp._dist if isinstance(var_exp, Int) or \
+                                  isinstance(var_exp, Float) else None \
+                                  for var_id, var_exp in self.exp.choices().items()}
+        else:
```

Comment:
in this code, either the user as 0 default distribution or it has to define all of them, the user is allowed to pass only a few distributions. Also, this code will not accept multi-variate distributions where a key can be a "tuple" of variable names.